plugin/core/transports.py

```diff
@@ -18,49 +17,100 @@
 import weakref
 
 
-class Transport(metaclass=ABCMeta):
+T = TypeVar('T')
+T_contra = TypeVar('T_contra', contravariant=True)
 
-    @abstractmethod
-    def send(self, payload: Dict[str, Any]) -> None:
-        pass
 
-    @abstractmethod
+class StopLoopError(Exception):
+    pass
+
+
+class Transport(Generic[T]):
+
+    def send(self, payload: T) -> None:
+        raise NotImplementedError()
+
     def close(self) -> None:
-        pass
+        raise NotImplementedError()
 
 
-class TransportCallbacks(Protocol):
+class TransportCallbacks(Protocol[T_contra]):
 
     def on_transport_close(self, exit_code: int, exception: Optional[Exception]) -> None:
         ...
 
-    def on_payload(self, payload: Dict[str, Any]) -> None:
+    def on_payload(self, payload: T_contra) -> None:
         ...
 
     def on_stderr_message(self, message: str) -> None:
         ...
 
 
-class JsonRpcTransport(Transport):
+class AbstractProcessor(Generic[T]):
+
+    def write_data(self, writer: IO[bytes], data: T) -> None:
+        raise NotImplementedError()
+
+    def read_data(self, reader: IO[bytes]) -> Optional[T]:
+        raise NotImplementedError()
+
+
+class JsonRpcProcessor(AbstractProcessor[Dict[str, Any]]):
+
+    def write_data(self, writer: IO[bytes], data: Dict[str, Any]) -> None:
+        body = self._encode(data)
+        writer.writelines(("Content-Length: {}\r\n\r\n".format(len(body)).encode('ascii'), body))
+
+    def read_data(self, reader: IO[bytes]) -> Optional[Dict[str, Any]]:
+        headers = http.client.parse_headers(reader)  # type: ignore
+        try:
+            body = reader.read(int(headers.get("Content-Length")))
+        except TypeError:
+            # Expected error on process stopping. Stop the read loop.
+            raise StopLoopError()
```

Comment:
Change so that explicit action is used to stop the read loop instead of relying on `TypeError` doing it.

We could maybe just raise `StopLoopError()` earlier by checking if `headers.get("Content-Length")` is `None`.